### PR TITLE
BUG: Use self._constructor_sliced in df._reduce to respect subclassed…

### LINKED ISSUE #25596: pandas.DataFrame.sum() returns wrong type for subclassed pandas DataFrame
#### Code Sample, a copy-pastable example if possible

```python

# the following code is obtained from the documentation
# https://pandas.pydata.org/pandas-docs/stable/development/extending.html

import pandas as pd

class SubclassedSeries(pd.Series):
    @property
    def _constructor(self):
        return SubclassedSeries
    @property
    def _constructor_expanddim(self):
        return SubclassedDataFrame


class SubclassedDataFrame(pd.DataFrame):
    @property
    def _constructor(self):
        return SubclassedDataFrame
    @property
    def _constructor_sliced(self):
        return SubclassedSeries

# create a class instance as in the example of the documentation

df = SubclassedDataFrame({'A': [1, 2, 3], 'B': [4, 5, 6], 'C': [7, 8, 9]})
>>> df
   A  B  C
0  1  4  7
1  2  5  8
2  3  6  9

# this works just fine

>>> type(df)
<class '__main__.SubclassedDataFrame'>

# slicing also works fine

>>> sliced2 = df['A']
>>> sliced2
0    1
1    2
2    3
Name: A, dtype: int64

>>> type(sliced2)
<class '__main__.SubclassedSeries'>

# however, the sum operation returns a pandas.Series, not SubclassedSeries

>>> sliced3 = df.sum()
>>> sliced3
0    1
1    2
2    3
Name: A, dtype: int64

>>> type(sliced3)
<class 'pandas.core.series.Series'>

```
#### Problem description

In our project, we extend pandas as described in the documentation and implement our own kind of DataFrame and Series, similar to the geopandas project (if you apply sum on their DataFrame, the same problem appears). If you want to use _reduce operations like sum, it is important that the correct SubclassedSeries is returned. Otherwise, inheritance from pandas.DataFrames is not possible.

#### Expected Output
```python
>>> type(sliced3)
<class '__main__.SubclassedSeries'>

```

I think I can provide a possible fix of this problem: The relevant code is contained in core/frame.py just before the return statement of the _reduce function:

```python

# this is the code in core/frame.py:
def _reduce(...):
        # .... left out
        if constructor is not None:
            result = Series(result, index=labels)
        return result

# I suggest the following change:
def _reduce(...):
        # .... left out
        if constructor is None:
            result = Series(result, index=labels)
        else:
            result = constructor(result, index=labels)
        # alternative (since constructor will create a SubclassedDataFrame):
            result = self._constructor_sliced(result, index=labels)
        return result

```

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.2.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.1
pytest: None
pip: 19.0.3
setuptools: 40.8.0
Cython: None
numpy: 1.16.2
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by EmilianoJordan**
… types. (GH25596)

- [x] closes #25596
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @EmilianoJordan for the PR. will need whatsnew once #30907 is merged.

File: pandas/tests/frame/test_subclass.py [hunk 1]
```diff
@@ -557,3 +557,26 @@ def strech(row):
         result = df.apply(lambda x: [1, 2, 3], axis=1)
         assert not isinstance(result, tm.SubclassedDataFrame)
         tm.assert_series_equal(result, expected)
+
+    def test_subclassed_df_reduce_functions(self):
```
Comment (simonjayhawkins) [inline pandas/tests/frame/test_subclass.py]: can this test be parametrized and split to use all_numeric_reductions and all_boolean_reductions fixtures from conftest.py

Comment (EmilianoJordan) [reply pandas/tests/frame/test_subclass.py]: Thanks for the feedback. This is done with the most recent commit. Waiting for the automated tests to mark this as resolved.

**Comment by EmilianoJordan** [comment]
> Thanks @EmilianoJordan for the PR. will need whatsnew once #30907 is merged.

Subscribed to #30907 to know when I can do this. Also, updated the initial pull request comment to reflect that this has not been completed.

File: pandas/tests/frame/test_subclass.py [hunk 2]
```diff
@@ -557,3 +557,25 @@ def strech(row):
         result = df.apply(lambda x: [1, 2, 3], axis=1)
         assert not isinstance(result, tm.SubclassedDataFrame)
         tm.assert_series_equal(result, expected)
+
+    def test_subclassed_numeric_reductions(self, all_numeric_reductions):
+        # GH 25596
+
+        op = all_numeric_reductions
+
+        df = tm.SubclassedDataFrame({"A": [1, 2, 3], "B": [4, 5, 6], "C": [7, 8, 9]})
```
Comment (jreback) [inline pandas/tests/frame/test_subclass.py]: can you remove the extra blank lines between statements.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. pls add a whatsnew note for 1.1, you can put in the bug fix section for reshaping.

**Comment by EmilianoJordan** [comment]
@jreback, @simonjayhawkins Quick question tangentially related to this. 

I noticed that DataFrame.count() should suffer from the same issue.
I also noticed that DataFrame.hdf() fails for subclassed types and I imagine that Series.to_hdf() will behave the same. 

Should I submit a bug report and then submit a pull request or is the community accepting of direct PRs?

Thanks, E

File: doc/source/whatsnew/v1.1.0.rst [hunk 3]
```diff
@@ -140,7 +140,7 @@ Groupby/resample/rolling
 Reshaping
 ^^^^^^^^^
 
--
+- Bug in :meth:`DataFrame._reduce` not returning subclassed data type effecting all numeric and boolean reduction methods. (:issue:`25596`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.1.0.rst]: this is not a public method, can you update with something a user would want to know, e.g. ``DataFrame.sum()`` etc.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.1.0.rst (hunk 3 above)
Comment (EmilianoJordan) [reply doc/source/whatsnew/v1.1.0.rst]: List all 12 methods then?

Comment (EmilianoJordan) [reply doc/source/whatsnew/v1.1.0.rst]: I'm working on a more comprehensive review of subclassing. How about I just add that to this PR and list all the methods? It'll be ~20 methods, but it'll get it all out of the way in one go.

Comment (jreback) [reply doc/source/whatsnew/v1.1.0.rst]: u can list if u want or just sat reduction methods

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @EmilianoJordan

**Pull request merged**